### ISSUE TITLE
Arrow keys and backspace not working in bash after recent update

### ISSUE THREAD

**Issue by avih** [msys2/MSYS2-packages#1124]
Today I ran `pacman -Syu` which upgraded the following packages (from the end of `/var/log/pacman.log`):
```
[2018-01-30 11:56] [ALPM] upgraded mpfr (3.1.5.1-3 -> 4.0.0.1-3)
[2018-01-30 11:56] [ALPM] upgraded gcc-libs (6.4.0-1 -> 6.4.0-2)
[2018-01-30 11:56] [ALPM] upgraded ncurses (6.0.20170708-2 -> 6.1.20180127-1)
[2018-01-30 11:56] [ALPM] upgraded gawk (4.1.4-2 -> 4.2.0-1)
[2018-01-30 11:56] [ALPM] upgraded isl (0.16.1-1 -> 0.18-1)
[2018-01-30 11:56] [ALPM] upgraded mpc (1.0.3-1 -> 1.1.0-1)
[2018-01-30 11:56] [ALPM] upgraded msys2-w32api-headers (5.0.0.4961.e09dd4ca-1 -> 6.0.0.5097.36d7b92b-1)
[2018-01-30 11:56] [ALPM] upgraded msys2-w32api-runtime (5.0.0.4961.e09dd4ca-1 -> 6.0.0.5097.36d7b92b-1)
[2018-01-30 11:56] [ALPM] upgraded gcc (6.4.0-1 -> 6.4.0-2)
[2018-01-30 11:56] [ALPM] upgraded ncurses-devel (6.0.20170708-2 -> 6.1.20180127-1)
[2018-01-30 11:56] [ALPM] upgraded wget (1.19.2-1 -> 1.19.3-1)
```

After this update the arrow keys and backspace don't work correctly at the shell input line, in a weird way.

I tried the following with and without a `~/.bashrc` file:

Pressing few chars and then a backspace seems to print a space. Up arrow works but then the shell doesn't clean up the line with history inputs and just appends the history items to the line.

Left and right arrows at the shell prompt don't skip chars as expected, but `ctrl-v` and then pressing an arrow key does print a correct (I think) string, e.g. ctrl-v and then right arrow prints `^[[C` but the shell doesn't seem to re-render the line correctly when used without it.

Up/down/left/right arrows while in `less` (/usr/bin/less) does seem to work correctly.

It's like the shell reads the keys correctly but not able to re-render the input line correctly. Maybe the ncurses update is at fault?

**Comment by avih** [comment]
It seems the shell is not able to move the cursor backwards, and when it tries to re-render the line, it just appends the new string instead of replacing the old one. This seems to happen with arrows and backspace.

E.g. typing `echo x<backspace>y` does print `y`, but shows at the command line while typing as `$ echo x y`, and typing `echo x<left-arrow>y` does print `yx` but shows at the input line as `$ echo xyx`.

**Comment by dandersson** [comment]
Just happened for me as well. Downgrading `ncurses` solved it for me.

I checked my `/var/log/pacman.log` which ended with:

```
[2018-01-15 09:48] [PACMAN] Running 'pacman -Syu'
[2018-01-15 09:48] [PACMAN] synchronizing package lists
[2018-01-15 09:48] [PACMAN] starting core system upgrade
[2018-01-15 09:49] [ALPM] transaction started
[2018-01-15 09:49] [ALPM] upgraded mintty (1~2.8.1-1 -> 1~2.8.3-1)
[2018-01-15 09:49] [ALPM] transaction completed
[2018-01-15 09:49] [PACMAN] Running 'pacman -Syu'
[2018-01-15 09:49] [PACMAN] synchronizing package lists
[2018-01-15 09:49] [PACMAN] starting core system upgrade
[2018-01-15 09:49] [PACMAN] starting full system upgrade
[2018-01-15 09:49] [ALPM] transaction started
[2018-01-15 09:49] [ALPM] removed tmux-git (2.5.94.g73b9328c-1)
[2018-01-15 09:49] [ALPM] upgraded gzip (1.8-1 -> 1.9-1)
[2018-01-15 09:49] [ALPM] installed tmux (2.6-1)
[2018-01-15 09:49] [ALPM] upgraded wget (1.19.1-3 -> 1.19.2-1)
[2018-01-15 09:49] [ALPM] transaction completed
[2018-01-15 09:49] [PACMAN] Running 'pacman -Syu'
[2018-01-15 09:49] [PACMAN] synchronizing package lists
[2018-01-15 09:49] [PACMAN] starting core system upgrade
[2018-01-15 09:49] [PACMAN] starting full system upgrade
[2018-01-30 11:17] [PACMAN] Running 'pacman -Syu'
[2018-01-30 11:17] [PACMAN] synchronizing package lists
[2018-01-30 11:17] [PACMAN] starting core system upgrade
[2018-01-30 11:17] [PACMAN] starting full system upgrade
[2018-01-30 11:19] [ALPM] transaction started
[2018-01-30 11:19] [ALPM] upgraded mpfr (3.1.5.1-3 -> 4.0.0.1-3)
[2018-01-30 11:19] [ALPM] upgraded gcc-libs (6.4.0-1 -> 6.4.0-2)
[2018-01-30 11:20] [ALPM] upgraded ncurses (6.0.20170708-2 -> 6.1.20180127-1)
[2018-01-30 11:20] [ALPM] upgraded gawk (4.1.4-2 -> 4.2.0-1)
[2018-01-30 11:20] [ALPM] upgraded isl (0.16.1-1 -> 0.18-1)
[2018-01-30 11:20] [ALPM] upgraded mpc (1.0.3-1 -> 1.1.0-1)
[2018-01-30 11:20] [ALPM] upgraded msys2-w32api-headers (5.0.0.4961.e09dd4ca-1 -> 6.0.0.5097.36d7b92b-1)
[2018-01-30 11:20] [ALPM] upgraded msys2-w32api-runtime (5.0.0.4961.e09dd4ca-1 -> 6.0.0.5097.36d7b92b-1)
[2018-01-30 11:20] [ALPM] upgraded gcc (6.4.0-1 -> 6.4.0-2)
[2018-01-30 11:20] [ALPM] upgraded openssh (7.5p1-1 -> 7.6p1-1)
[2018-01-30 11:20] [ALPM] upgraded git (2.15.0-1 -> 2.16.1-1)
[2018-01-30 11:20] [ALPM] upgraded tzcode (2017.b-1 -> 2018.c-1)
[2018-01-30 11:20] [ALPM] upgraded wget (1.19.2-1 -> 1.19.3-1)
[2018-01-30 11:20] [ALPM] transaction completed
```

So, somewhere between 2018-01-15 and 2018-01-30 this regression happened. `ncurses` seemed like the most suspicious upgrade, so I [downgraded](https://wiki.archlinux.org/index.php/downgrading_packages) through:

    pacman -U /var/cache/pacman/pkg/ncurses-6.0.20170708-2-x86_64.pkg.tar.xz

and in the next MSYS2 instance I started, terminal codes for backspace, delete, etc., worked as intended again.

I tried checking the `Backarrow sends ^H` and `Delete sends DEL` settings in "broken" mode, but it did not seem to help, instead doing nothing when pressing those characters, if I recall correctly.

I am using `TERM=xterm-256color`.

**Comment by Alexpux** [comment]
I'm not planning downgrade ncurses. Instead need to find why it happen

**Comment by avih** [comment]
FWIW, I tried rebuilding bash locally (first time, so hopefully it's OK) and it seems to fix the issue. Additionally, the release notes say:
> hese notes are for ncurses 6.1, released January 27, 2018.
>
> This release is designed to be source-compatible with ncurses 5.0 through 6.0; providing extensions to the application binary interface (ABI) ...

So perhaps all packages depending on ncurses need re-compile?

**Comment by Alexpux** [comment]
I cant reproduce issue on my computer. All keys seems work correctly for me

**Comment by avih** [comment]
FWIW, my original `/usr/bin/bash.exe` is from 2017-02-15, so almost a year old, and building a fresh one (same version `4.4.12(1)-release (x86_64-pc-msys)`, using makepkg) did fix it for me. I also know of other people bumping into this issue, which, again, is quite a dealbreaker.

**Comment by elieux** [comment]
I do have the described issues as well.  Changing `TERM` (e.g. to `ansi`) seems to fix everything, but it's weirder than that:

```
ether@jolteon MSYS /w
$ echo $TERM
xterm-256color

ether@jolteon MSYS /w
[broken rendering]

ether@jolteon MSYS /w
$ TERM=xterm-256color

ether@jolteon MSYS /w
[still broken rendering]

ether@jolteon MSYS /w
$ TERM=xterm

ether@jolteon MSYS /w
[rendering fine]

ether@jolteon MSYS /w
$ TERM=xterm-256color

ether@jolteon MSYS /w
[rendering still fine]
```

-- 
David Macek

**Comment by niklasholm** [comment]
@Alexpux , what `TERM` are you using?

I have this issue as well. I reinstalled MSYS2 from scratch and updated all packages, no issue so far. Then I set my `TERM` to `xterm-256color` which caused the issue to appear; downgrading ncurses to 6.0 fixes the issue. Clearly this is a regression in or triggered by ncurses 6.1; the changelog mentions some changes to `xterm-256color`.

**Comment by chhe** [comment]
I can confirm the error. But only on Windows 8.1 using TERM=screen-256color or TERM=xterm-256color. On Windows 10 (Fall Creators Update) both TERMs don't produce this error (at least for me).

**Comment by lazka** [comment]
Deleting ~/.minttyrc helped here (or I was just lucky and that was unrelated..)

**Comment by avih** [comment]
> Deleting ~/.minttyrc helped here (or I was just lucky and that was unrelated..)

It's because the default mintty TERM is `xterm`, and the issue appears to be related to TERM being `xterm-256color` - which is required for some programs to display more than 8 colors.

**Comment by zklhp** [comment]
Rebuild bash can fix the problem. Thank you for the information.

You could try the rebuilded bash at

https://ci.appveyor.com/project/zklhp/msys2-packages/build/1.0.16/artifacts

**Comment by lazka** [comment]
#1128

**Comment by avih** [comment]
> #1128

Assuming ncurses 6.1 may break binary compatibility, it would need a rebuild of all packages which use ncurses. Bash is just the most visible of those (assuming there are more packages depending on ncurses).

**Comment by lazka** [comment]
maybe related: https://bugs.archlinux.org/task/57300?project=1&string=ncurses

**Comment by elieux** [comment]
Possibly interesting development: https://github.com/mintty/mintty/issues/738#issuecomment-364402557

-- 
David Macek

**Comment by avih** [comment]
> Possibly interesting development: mintty/mintty#738 (comment)

That link lacks info, but void linux appears to have had a similar issue: https://forum.voidlinux.eu/t/solved-problems-after-todays-update/4553 . Apparently it was fixed for them after bash was updated, but I don't know it it was just rebuilt or maybe something else.

Additionally, I now can't build bash with makepkg, there's some compile error with an unknown type IIRC - this is after the msys2 runtime was updated to cygwin 2.10 .

**Comment by avih** [comment]
Also, there appear to be two issues:
1. (fixed after rebuilding bash): editing keys don't work in CLI in bash with `TERM=xterm-256color` .
2. (NOT fixed after rebuilding bash): bash aborts after setting TERM to some dummy name, e.g. `TERM=blahblah` .

> I just rebuilt bash from current msys2-packages and installed it and the issue seems gone.

Would you mind sharing your build procedure? I cd to the bash dir of the msys2 packages repo and issued `makepkg --skippgpchec -sCLf` and it stopped/errored after some time with IIRC undefined type or something.

**Comment by elieux** [comment]
>  2. (NOT fixed after rebuilding bash): bash aborts after setting TERM to some dummy name, e.g. |TERM=blahblah| .

Okay, yeah, I didn't know of this issue. My rebuilt bash still does this.

> Would you mind sharing your build procedure? I cd to the bash dir of the msys2 packages repo and issued |makepkg --skippgpchec -sCLf| and it stopped/errored after some time with IIRC undefined type or something.

I did nothing special.  Just updated the repo and simple `makepkg` in the bash directory.

-- 
David Macek

**Comment by revelator** [comment]
Just to chime in the new ssp changes break building ncurses, undefined references to FILE in ssp/stdio.h and ssp/whar.h.

So guess it might need some patches to get around this.

**Comment by revelator** [comment]
Uh oh new msys2-runtime breaks windows cmake, it cannot find itself after update.
Reinstalling does not work either and i cannot recompile it either because the bootstrapped version also cannot find itself. Something tells me we should hold off on this version untill it becomes more stable.

**Comment by revelator** [comment]
for the actual error ill post it below.

´´´
-- The C compiler identification is GNU 7.3.0
-- The CXX compiler identification is GNU 7.3.0
-- Check for working C compiler: C:/Msys64/mingw64/bin/gcc.exe
-- Check for working C compiler: C:/Msys64/mingw64/bin/gcc.exe -- broken
CMake Error at Modules/CMakeTestCCompiler.cmake:52 (message):
  The C compiler

    "C:/Msys64/mingw64/bin/gcc.exe"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: C:/Msys64/src/MINGW-packages/mingw-w64-cmake/src/build-x86_64-w64-mingw32/CMakeFiles/CMakeTmp

    Run Build Command:"C:/Msys64/usr/bin/make.exe" "cmTC_20e2c/fast"
    gmake -f CMakeFiles/cmTC_20e2c.dir/build.make CMakeFiles/cmTC_20e2c.dir/build
    gmake[1]: G+▒r til katalog "/src/MINGW-packages/mingw-w64-cmake/src/build-x86_64-w64-mingw32/CMakeFiles/CMakeTmp"
    /bin/sh: /C/Msys64/src/MINGW-packages/mingw-w64-cmake/src/build-x86_64-w64-mingw32/Bootstrap.cmk/cmake.exe: No such file or directory
    CMakeFiles/cmTC_20e2c.dir/build.make:64: *** [CMakeFiles/cmTC_20e2c.dir/testCCompiler.c.obj] Error 127
    CMakeFiles/cmTC_20e2c.dir/build.make:65: recipe for target 'CMakeFiles/cmTC_20e2c.dir/testCCompiler.c.obj' failed

    #0  CMakeFiles/cmTC_20e2c.dir/testCCompiler.c.obj at /src/MINGW-packages/mingw-w64-cmake/src/build-x86_64-w64-mingw32/CMakeFiles/CMakeTmp/CMakeFiles/cmTC_20e2c.dir/build.make:64
    #1  cmTC_20e2c.exe at /src/MINGW-packages/mingw-w64-cmake/src/build-x86_64-w64-mingw32/CMakeFiles/CMakeTmp/CMakeFiles/cmTC_20e2c.dir/build.make:95
    #2  CMakeFiles/cmTC_20e2c.dir/build at /src/MINGW-packages/mingw-w64-cmake/src/build-x86_64-w64-mingw32/CMakeFiles/CMakeTmp/CMakeFiles/cmTC_20e2c.dir/build.make:102
    gmake[1]: Forlader katalog "/src/MINGW-packages/mingw-w64-cmake/src/build-x86_64-w64-mingw32/CMakeFiles/CMakeTmp"
    make: *** [Makefile:126: cmTC_20e2c/fast] Error 2




  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:5 (project)


-- Configuring incomplete, errors occurred!
See also "C:/Msys64/src/MINGW-packages/mingw-w64-cmake/src/build-x86_64-w64-mingw32/CMakeFiles/CMakeOutput.log".
See also "C:/Msys64/src/MINGW-packages/mingw-w64-cmake/src/build-x86_64-w64-mingw32/CMakeFiles/CMakeError.log".
---------------------------------------------
Error when bootstrapping CMake:
Problem while running initial CMake
---------------------------------------------
´´´

**Comment by revelator** [comment]
Hmm ok scratch that, for some reason one of the updates had deleted the fstab file, after recreating it it worked again. I noticed few missing files in msys2 filesystem package like the one that autoregenerates fstab, what was the reason this one was removed ? i newer had any problem with it so im curious.

**Comment by oldmud0** [comment]
I start my mintty session with `xterm-256color`, and I can reproduce the backspace problem as stated. When I enter `TERM=xterm`, the problem disappears, even after I set `TERM=xterm-256color` again.

I'll wait patiently for a release to get to pacman.

**Comment by LigH-de** [comment]
The update for bash to version 4.4.019-1 fixed the shell history for me.

**Comment by oldmud0** [comment]
Yep! Fixed it for me too.

**Comment by avih** [comment]
Bash 4.4.0.19 fixes the editing key issues for me with xterm-256color, but this new issue remains (both with 4.4.0.12 after a rebuild and also with the pre-built 4.4.0.19 after the update):

> (Only after rebuilding bash): bash aborts after setting TERM to some dummy name, e.g. TERM=blahblah

**Comment by revelator** [comment]
Fixed it for me rebuilding ncurses against the new cygwin ssp allthough i had to hack about it a bit because the cygwin ssp bailed out with an error to find struct FILE.  It works in other packages though so id say this is something that needs to be looked at upstream by the ncurses devs. 
One way to get around it is by including stdio.h before any of the ssp headers in ncurses.

**Comment by Alexpux** [comment]
Problem with finding FILE strict is solved yesterday. You need remove -D_FORTIFY_SOURCE=2 from CPPFLAGS in makepkg.conf file

**Comment by revelator** [comment]
Ok thats a bit simpler, but wont it affect other sources to remove the hardening ?

**Comment by Wybxc** [comment]
I solved the problem by using `ash.exe` and run `pacman -Syu bash` in it.

**Comment by elieux** [comment]
The PR linked just above should work around the crash, I hope.

**Comment by avih** [comment]
> The PR linked just above should work around the crash, I hope.

Confirm it fixes the crash for me, thanks. The cursor keys work too since the bash 4.4.0.19-1 and still with 4.4.0.19-2 .

**Comment by elieux** [comment]
@avih, good.

@Alexpux, what's wrong with `_FORTIFY_SOURCE=2`?  Isn't there a better fix for `struct FILE`?

**Comment by Alexpux** [comment]
@elieux there are conflicts in cygwin headers

**Comment by lazka** [comment]
From what I understand this is fixed, so closing.
